uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -379,9 +455,14 @@ define([
      */
     function initListeners () {
       listeners.push(
-        $rootScope.$on('LeaveRequestPopup::updateBalance', function () {
-          vm.updateBalance();
-          setDaySelectionMode();
+        $rootScope.$on('LeaveRequestPopup::updateBalance', function (event, absenceTypes) {
+          var previousAbsenceTypeId = vm.selectedAbsenceType.id;
+          if (absenceTypes) {
+            vm.absenceTypes = absenceTypes;
+          }
+          setSelectedAbsenceType();
```

Comment:
again i can't stress enough how really hard is to make sense of all of this if you are not the author of this code. Here's my understanding of the whole thing

1. I receive an event about the balance being updated
2. as part of the event about the balance, i get a bunch of absence types which i know nothing about (why are they passed? Do they have something special compared than the absence types are already there?)...
3. ...but not always, so i need to check if they are there
4. In any case, i store away the currently selected absence type id
5. Then I call a `setSelectedAbsenceType` function, which is really just a call to another function...
    ```js
    function setSelectedAbsenceType () {
      vm.selectedAbsenceType = getSelectedAbsenceType();
    }
    ```
6. ...whose job is to find the absence type that the leave request is currently assigned to, which is then assigned as the currently selected absence type (`vm.selectedAbsenceType`)
7. Both functions, although very generic in nature are never used again, meaning that they were created specifically for this event *about the balance*
8. In any case if the previously selected absence type differs from the currently selected absence type (why would that happen? Why would it happen when we update the balance, specifically? No idea)
    ```js
    (previousAbsenceTypeId !== vm.selectedAbsenceType.id)
    ````
    then I do something about the days selection mode (again, why?)
    ```js
    setDaysSelectionMode()
    ```
9. Only at the end I do something that finally has something to do with the event itself, that is, attempting at calculating the balance!
    ```js
    attemptCalculateBalanceChange()
    ```

I hope that this gave you a "fresh eyes" point of view, and why this part is in desperate need of refactoring

Reply:
I attempted to refactor it and separated into 2 events.